src/blockchain/blockchain_monitor.ts

```diff
@@ -0,0 +1,139 @@
+    // TODO: raise component:error event if monitor throws exception
+    this.#taskHandle = setInterval(() => this.#monitor(), this.#pollIntervalMs);
+
+    return super.start();
+  }
+
+  stop(): Promise<void> {
+    clearInterval(this.#taskHandle);
+
+    return super.stop();
+  }
+
+  async #monitor() {
+    this.logger.debug("Gathering blockchain state");
+
+    const mempool = [];
+    // the loop following this where we go through all the mempool txs could
+    // go in here as well but each time the `monitor:mempool-tx` event is raised
+    // it could provide a different mempool snapshot to plugins.
+    //
+    // instead, collect the full mempool first so the plugins can receive a full consistent snapshot.
+    for await (const page of this.#blockchainClient.getMempool()) {
+      mempool.push(...page);
+    }
+
+    for (const tx of mempool) {
+      if (!this.#state.mempoolTxDelivery[tx.id]) {
+        this.#state.mempoolTxDelivery[tx.id] = true;
+
+        this.dispatchEvent(
+          new CustomEvent("monitor:mempool-tx", { detail: tx }),
+        );
+      }
+
+      this.#state.pastMempoolTxIds = this.#state.pastMempoolTxIds.filter((
+        txId,
+      ) => txId !== tx.id);
+
+      // remove txid from undefined state transactions map if present
+      // this resets the mempool drop detection counter for this txid
+      delete this.#state.mempoolTxChecks[tx.id];
+    }
+
+    // if tx was present in previous mempool, but not in the
+    // current, it may have been dropped or included in a block
+    for (const txId of this.#state.pastMempoolTxIds) {
+      this.#state.mempoolTxChecks[txId] =
+        (this.#state.mempoolTxChecks[txId] ?? 0) + 1;
+    }
+
+    this.#state.pastMempoolTxIds = mempool.map((tx) => tx.id);
+
+    const currentHeight = await this.#blockchainClient.getCurrentHeight();
+
+    if (currentHeight > this.#state.currentHeight) {
+      const newBlock = await this.#blockchainClient.getBlock(
+        currentHeight,
+      ) as any;
+
+      this.dispatchEvent(
+        new CustomEvent("monitor:new-block", { detail: newBlock }),
+      );
+
+      this.#state.currentHeight = currentHeight;
+
+      for (const tx of newBlock.blockTransactions) {
+        this.dispatchEvent(
+          new CustomEvent("monitor:included-tx", { detail: tx }),
+        );
+
+        // stop tracking mempool delivery for this txid
+        delete this.#state.mempoolTxDelivery[tx.id];
+
+        // prevent `onMempoolTxDrop` event for this txid as
+        // it is now included in a block
+        delete this.#state.mempoolTxChecks[tx.id];
+      }
+    }
+
+    for (const txId of Object.keys(this.#state.mempoolTxChecks)) {
+      // if a tx is not included in a block in dropChecks * `n` seconds,
+      // then it's probably dropped from the mempool
+      if (this.#state.mempoolTxChecks[txId] > this.#maxMempoolTxChecks) {
+        // TODO raise mempool dropped, so we still need to keep track of txns not just the id?
```

Comment:
@arobsn when a tx is dropped from mempool should we include the transaction when raising the event for plugins or would txId be enough?

Reply:
Good question, I would go with the full transaction to keep hooks homogeneous and avoid more code on plugins side (IMO plugins must be as concise as possible to make auditing process easier).